/*
* @license Apache-2.0
*
* Copyright (c) 2019 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/* tslint:disable:no-construct */

import forOwn = require( './index' );

const log = ( v: any, key: string ): string => {
	return `${key}: ${v}`;
};

// TESTS //

// The function returns an object...
{
	forOwn( new String( 'abc' ), log ); // $ExpectType any
	forOwn( new String( 'abc' ), log, {} ); // $ExpectType any
}

// The compiler throws an error if the function is provided a second argument which is not a function...
{
	forOwn( new String( 'abc' ), 2 ); // $ExpectError
	forOwn( new String( 'abc' ), false ); // $ExpectError
	forOwn( new String( 'abc' ), true ); // $ExpectError
	forOwn( new String( 'abc' ), 'abc' ); // $ExpectError
	forOwn( new String( 'abc' ), {} ); // $ExpectError
	forOwn( new String( 'abc' ), [] ); // $ExpectError
}

// The compiler throws an error if the function is provided an invalid number of arguments...
{
	forOwn(); // $ExpectError
	forOwn( new String( 'abc' ) ); // $ExpectError
	forOwn( new String( 'abc' ), log, {}, 3 ); // $ExpectError
}
