// @sentry/core/esm/index.js
export { addBreadcrumb, captureEvent, captureException, captureMessage, configureScope, setContext, setExtra, setExtras, setTag, setTags, setUser, startTransaction, withScope } from './exports.js';
export { Hub, getCurrentHub, getHubFromCarrier, getMainCarrier, makeMain, setHubOnCarrier } from './hub.js';
export { closeSession, makeSession, updateSession } from './session.js';
export { SessionFlusher } from './sessionflusher.js';
export { Scope, addGlobalEventProcessor } from './scope.js';
export { getEnvelopeEndpointWithUrlEncodedAuth, getReportDialogEndpoint } from './api.js';
export { BaseClient } from './baseclient.js';
export { initAndBind } from './sdk.js';
export { createTransport } from './transports/base.js';
export { makeOfflineTransport } from './transports/offline.js';
export { SDK_VERSION } from './version.js';
export { getIntegrationsToSetup } from './integration.js';
import * as index from './integrations/index.js';
export { index as Integrations };
export { prepareEvent } from './utils/prepareEvent.js';
export { FunctionToString } from './integrations/functiontostring.js';
export { InboundFilters } from './integrations/inboundfilters.js';
